### DISCUSSION TITLE
Any way to deal with broken package on otherwise useful repo?

### DISCUSSION THREAD

**Discussion by ppanon2022** [uyuni-project/uyuni#7549] [General]
Microsoft has a number of useful packages on their product repo at
https://packages.microsoft.com/ubuntu/20.04/prod/dists/focal/main/binary-amd64/
Unfortunately, that repo also contains 10 broken packages missing files which cause a sync failure notification for every sync attempt (other packages on that repo sync OK).
Microsoft are unlikely to fix those packages retroactively. Is there any other way to avoid those downloads/import and/or notifications somehow without disabling all sync failure notifications?

2023/08/30 03:12:48 -07:00   Downloading packages:
2023/08/30 03:12:49 -07:00     1/10 : intune-portal_1.2210.50_amd64.deb
2023/08/30 03:12:51 -07:00     2/10 : intune-portal_1.2302.11_amd64.deb
2023/08/30 03:12:51 -07:00     3/10 : intune-portal_1.2211.21_amd64.deb
2023/08/30 03:12:51 -07:00     4/10 : intune-portal_1.2302.9_amd64.deb
2023/08/30 03:12:51 -07:00     5/10 : intune-portal_1.2307.12_amd64.deb
2023/08/30 03:12:51 -07:00     6/10 : intune-portal_1.2305.20_amd64.deb
2023/08/30 03:12:51 -07:00     7/10 : libodbc1_2.3.7_amd64.deb
2023/08/30 03:12:52 -07:00     8/10 : odbcinst1debian2_2.3.7_amd64.deb
2023/08/30 03:12:52 -07:00     9/10 : intune-portal_1.2303.10_amd64.deb
2023/08/30 03:13:03 -07:00     10/10 : msopenjdk-11_11.0.20.1-1_amd64.deb
2023/08/30 03:13:03 -07:00 Importing packages started.
2023/08/30 03:13:03 -07:00 
2023/08/30 03:13:03 -07:00   Importing packages to DB:
2023/08/30 03:13:03 -07:00 File not found inside package
2023/08/30 03:13:03 -07:00 File not found inside package
2023/08/30 03:13:03 -07:00 File not found inside package
2023/08/30 03:13:03 -07:00 File not found inside package
2023/08/30 03:13:03 -07:00 File not found inside package
2023/08/30 03:13:03 -07:00 File not found inside package
2023/08/30 03:13:03 -07:00 File not found inside package
2023/08/30 03:13:03 -07:00 File not found inside package
2023/08/30 03:13:04 -07:00 File not found inside package
2023/08/30 03:13:05 -07:00   Package batch #1 of 1 completed...

Cheers,

Paul-Andre

**Comment by juliogonzalez** [comment]
What is the error? The reposync crashes at `2023/08/30 03:13:05 -07:00 Package batch https://github.com/uyuni-project/uyuni/pull/1 of 1 completed...` without showing any error?

**Reply by ppanon2022** [reply]
It's not an error exactly and the reposync doesn't crash. Because those packages are poorly constructed, those ten instances of 
2023/08/30 03:13:04 -07:00 File not found inside package
are resulting in a notification/alert every time the sync is run. It can never sync/upload those packages successfully so it retries anew every time. It's mainly just an admin nuisance, which can cause habituation/admin fatigue and overlooking a more significant error. I'm just wondering if there's any way to suppress the sync attempt/standards compliance test for those specific packages going forward. There are newer correct versions for most of those packages which have been imported OK (the intune-portal packages are the exception), so installing or updating to the newest release of the packages is still possible.

I'm wondering if it might be possible to generate dummy entries in the database so that reposync doesn't constantly try to alert on them. Initially I was thinking manual creating those DB entries, but I guess it's likely reposync does its file compliance check before checking if the package exists for that channel in Uyuni. But it would be nice if there was a way to run reposync to mark those packages as blacklisted so that any package compliance check failures on those specific packages are ignored.

**Reply by juliogonzalez** [reply]
I'd not recommend playing around with the database. There are other less dangerous workarounds that could work.

One I can think of is adding a `filter` for the repository. On the WebUI you can find that at Software > Manage > Repositories > Your repository > Detail. I think that if you add expressions starting with `-` that includes not just the name of the package but the version as well (and maybe architecture), for example `-omi-1.6.10-2.ssl_110.ulinux.x64`, that will work.

For RPM, this filter could be ignored if other packages require those packages in the filter. For DEB it seems it's not the case, according to @vzhestkov 

Otherwise, you can force reposync to only sync the latest packages. However that's managed on a channel/subchannel level. So you may need a single channel for this repository. You will find this option at Software > Manage > Channels > Your Channel > Repositories > Sync, then disable the scheduling, enable `sync only latest packages` and readd the scheduling. This should work according to @cbbayburt 

Be aware, however, that this could also mean some packages will become broken, as some packages on the last version could depend on other dependencies on a version that's not the latest one.

**Comment by ppanon2022** [comment]
Thank you very much Julio. I had some challenges getting the filter right, especially since I was confused because the error seems to still be in the sync log but doesn't generate a notification alert or email. But I seem to have successfully tweaked it to the point where the alerts and emails have stopped.

**Reply by juliogonzalez** [reply]
You are welcome.

Would you mind sharing the resulting filter, as workaround (at least for now) for other uses?

**Comment by ppanon2022** [comment]
The filter I settled on was 
-odbcinst1debian2_2.3.7_amd64*,-libodbc1_2.3.7_amd64*,-intune-portal*
